hudi-common/src/main/java/org/apache/hudi/stats/ValueType.java

```diff
@@ -0,0 +1,668 @@
+      case STRING:
+        if (schema.getLogicalType() == null) {
+          return ValueType.STRING;
+        } else if (Objects.equals(schema.getLogicalType().getName(), LogicalTypes.uuid().getName())) {
+          return ValueType.UUID;
+        }
+        throw new IllegalArgumentException("Unsupported logical type for String: " + schema.getLogicalType());
+      case FIXED:
+        if (schema.getLogicalType() == null) {
+          return ValueType.FIXED;
+        } else if (schema.getLogicalType() instanceof LogicalTypes.Decimal) {
+          return ValueType.DECIMAL;
+        }
+        throw new IllegalArgumentException("Unsupported logical type for Fixed: " + schema.getLogicalType());
+      case UNION:
+        return fromSchema(AvroSchemaUtils.resolveNullableSchema(schema));
+      default:
+        throw new IllegalArgumentException("Unsupported type: " + schema.getType());
+    }
+  }
+
+  // Casting to standardize types and also type promotion
+  private static Comparable<?> passThrough(Object val, ValueMetadata meta) {
+    return (Comparable<?>) val;
+  }
+
+  private static Boolean castToBoolean(Object val) {
+    if (val instanceof Boolean) {
+      return (Boolean) val;
+    } else {
+      throw new UnsupportedOperationException("Unable to convert boolean: " + val.getClass());
+    }
+  }
+
+  private static Integer castToInteger(Object val) {
+    if (val == null) {
+      return null;
+    }
+    if (val instanceof Integer) {
+      return (Integer) val;
+    } else if (val instanceof Long) {
+      return ((Long) val).intValue();
+    } else if (val instanceof Float) {
+      return ((Float) val).intValue();
+    } else if (val instanceof Double) {
+      return ((Double) val).intValue();
+    } else if (val instanceof Boolean) {
+      return ((Boolean) val) ? 1 : 0;
+    } else {
+      // best effort casting
+      return Integer.parseInt(val.toString());
+    }
+  }
+
+  private static Long castToLong(Object val) {
+    if (val == null) {
+      return null;
+    }
+    if (val instanceof Integer) {
+      return ((Integer) val).longValue();
+    } else if (val instanceof Long) {
+      return ((Long) val);
+    } else if (val instanceof Float) {
+      return ((Float) val).longValue();
+    } else if (val instanceof Double) {
+      return ((Double) val).longValue();
+    } else if (val instanceof Boolean) {
+      return ((Boolean) val) ? 1L : 0L;
+    } else {
+      // best effort casting
+      return Long.parseLong(val.toString());
+    }
+  }
+
+  private static Float castToFloat(Object val) {
+    if (val == null) {
+      return null;
+    }
+    if (val instanceof Integer) {
+      return ((Integer) val).floatValue();
+    } else if (val instanceof Long) {
+      return ((Long) val).floatValue();
+    } else if (val instanceof Float) {
+      return (Float) val;
+    } else if (val instanceof Double) {
+      return ((Double) val).floatValue();
```

Comment:
I remember there's particular issue around the precision on the conversion between double and float.  Is that handled here?

Reply:
Yeah that's covered. That's from float -> double. This is double -> float, which actually maybe I should remove? I got this from HoodieTableMetadataUtil.coerceToComparable

Reply:
Got it.  It should be removed if such evolution is not allowed.